lib/_stream_readable.js

```diff
@@ -1070,13 +1072,84 @@ Readable.prototype.wrap = function(stream) {
 };
 
 Readable.prototype[SymbolAsyncIterator] = function() {
-  if (createReadableStreamAsyncIterator === undefined) {
-    createReadableStreamAsyncIterator =
-      require('internal/streams/async_iterator');
+  let stream = this;
+
+  if (typeof stream.read !== 'function') {
+    // v1 stream
+    const src = stream;
+    stream = new Readable({
+      objectMode: true,
+      destroy(err, callback) {
+        destroyImpl.destroyer(src, err);
+        callback();
+      }
+    }).wrap(src);
   }
-  return createReadableStreamAsyncIterator(this);
+
+  const iter = createAsyncIterator(stream);
+  iter.stream = stream;
+  return iter;
 };
 
+async function* createAsyncIterator(stream) {
+  const state = stream._readableState;
+
+  if ((state && state.endEmitted) ||
+      (state && state.closeEmitted)) {
+    // TODO (ronag): What if errorEmitted?
+    // TODO (ronag): Wait for closeEmitted?
+    // TODO (ronag): if (closeEmitted && !endEmitted) ERR_PREMATURE_CLOSE?
+    stream.on('error', nop);
+    return;
+  }
+
+  let callback = nop;
+  let error;
+
+  stream
+    .on('readable', () => {
+      callback();
+      callback = nop;
+    })
+    .on('error', (err) => {
+      error = err;
+      callback();
+      callback = nop;
+    })
+    .on('end', () => {
+      error = error || null;
+      callback();
+      callback = nop;
+    })
+    .on('close', () => {
+      error = error || null;
+      // TODO (ronag): ERR_PREMATURE_CLOSE?
+      callback();
+      callback = nop;
+    });
```

Comment:
We could also put callback and error as symbols on the stream object. That way we can avoid the closures... not sure though if that is a pattern we encourage.